### PR TITLE
✨ Flat message array for code mode

### PR DESCRIPTION
**Pull request by TechNickAI**
## Summary

Implements a flat message array approach for code mode, based on competitive analysis of Claude Code web UIs (claudecodeui, claude-code-webui, opcode). All successful implementations use this pattern.

- **Flat message array** - Single array of messages, no nested threading
- **Tool result linking by ID** - Results update tool messages in-place, not separate messages  
- **Append-only mutations** - Only mutation is adding tool results to existing tools
- **Backwards compatible** - Context derives legacy format for existing rendering

### New: MessageProcessor (`lib/code/messages.ts`)
- Simple flat message array (`CodeMessage[]`)
- Four message types: `text`, `tool`, `user`, `system`
- Tool states: `streaming → running → complete/error`
- 29 tests covering real Claude Code output patterns

### Updated: Streaming Endpoint (`app/api/code/route.ts`)
- Uses `MessageProcessor` instead of `ToolStateAccumulator`
- Emits `data-code-messages` with flat array (simpler than legacy format)

### Updated: Context (`lib/code/tool-state-context.tsx`)
- Handles new `data-code-messages` format
- Derives legacy format (`tools`, `contentOrder`, `textSegments`) for backwards compatibility
- Adds `useCodeMessages()` hook for direct flat array access

## Bug Fixes (from code review)
- Use counter instead of `Date.now()` for message IDs (prevents collision in rapid calls)
- Update `toolName` on existing tool in `onToolCall` (handles placeholder names)

## Code Review Summary

Three automated review agents analyzed the code:

**Logic Review:** Found and fixed ID collision bug and toolName update bug. Verified state transitions are monotonic, backwards compatibility is correct, no race conditions.

**Style Review:** All code follows project conventions - import organization, naming, TypeScript types, structured logging, documentation.

**Test Review:** 29 tests cover real-world patterns well. Suggested additional edge case tests for future hardening.

## Test Plan
- [x] All 1570 tests pass
- [x] Type check passes
- [x] Lint passes
- [ ] Manual test with real Claude Code output

🤖 Generated with [Claude Code](https://claude.com/claude-code)

----

> [!IMPORTANT]
> Introduces a flat message array for code mode with `MessageProcessor`, updating `route.ts` and `tool-state-context.tsx` for new message handling while maintaining backward compatibility.
> 
>   - **Behavior**:
>     - Implements flat message array in `messages.ts` with `MessageProcessor` handling `text`, `tool`, `user`, and `system` messages.
>     - Updates `route.ts` to use `MessageProcessor` for streaming code agent responses.
>     - Maintains backward compatibility by deriving legacy format in `tool-state-context.tsx`.
>   - **Tests**:
>     - Adds 29 tests in `messages.test.ts` covering various message patterns and edge cases.
>   - **Misc**:
>     - Updates `tool-state-context.tsx` to handle new `data-code-messages` format and derive legacy format.
>     - Removes `ToolStateAccumulator` from `route.ts` and replaces it with `MessageProcessor`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=carmentacollective%2Fcarmenta&utm_source=github&utm_medium=referral)<sup> for 81f406268132c2f709dc677c7598d91910fee134. You can [customize](https://app.ellipsis.dev/carmentacollective/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by TechNickAI** [comment]
Thanks for the thorough review! Addressed your feedback:

**#1 Memory cleanup** - ✅ Fixed in ea6f8c0. Now delete `inputBuffers` entry when tool completes.

**#2 Unused imports** - These prepare for the next phase (direct flat array rendering). Keeping them to avoid import churn.

**#3 Type assertion** - Current pattern works reliably. TypeScript's narrowing behavior varies enough that the predicate form isn't guaranteed to be better.

**#4 State machine docs** - The type definitions (`CodeToolState = "streaming" | "running" | "complete" | "error"`) plus the 29 tests document the flow clearly. Additional comment would add noise without clarity.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: lib/code/messages.ts [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    addSystemMessage(content: string, isError = false): SystemMessage {
+        const message: SystemMessage = {
+            type: "system",
+            id: `system-${this.messageIndex++}`,
+            content,
+            isError,
+        };
+        this.messages.push(message);
+        return message;
+    }
```
[the remark this answers is not in the document]
Comment (TechNickAI) [reply lib/code/messages.ts]: Excellent find! Fixed in 81f4062 - now call `finalizeText()` and set `currentTextId = null` in `addSystemMessage`.
